### ISSUE TITLE
feat: use an LLM to generate synthetic sentence pairs

### ISSUE THREAD

**Issue by tom9358** [tom9358/nllb-tryout#34]
I've written liek 7000 sentences by hand now and the NLLB translator is improving, but not that rapidly. I think some kind of synthetic data is really needed if we ever want to make serious progress on this.

Fortunately, there are now LLMs that allow usage for generation of datasets, e.g. Qwen and Gemma models. While they're pretty bad at translation Dutch to Gronings, they can do translation Gronings to Dutch pretty well is my experience (from just trying it out a bit).

Let's try making a locally running gemma/qwen model translate Gronings texts to Dutch, and calculate chrf and bleu and exact matches. Then let's try to optimize the generation setup a little bit and see if we can get in the ballpark of valuable synthetic training data. Let's use Kreuze as a dataset, as the texts are already freely available online and the documents explicitly state: "Vrij rond te sturen noar elkenain dij t lezen wil" (i.e. free to share with anyone who wants to read it). I will preserve the source and attribution information when generating the synthetic data.

- [x] Test Qwen and Gemma on existing Tatoeba data
  - [x] Try a few instruction formulation variations
  - [x] Look into few-shot prompting
- [x] If results look promising, test translating longer text blocks. LLMs can use the larger text as context for better translations.
- [x] If this seems to work, generate synthetic data from Kreuze
- [x] Inspect whether the generated data is useful for training

To be clear, I will not train an LLM on this data. The goal is to create good quality Gronings machine translation.

**Comment by tom9358** [comment]
# Cost estimate

Doing this locally saves a lot of money compared to API of one of the big tech companies.
Assume:
- 300 Watt GPU, 100 Watt CPU and 100 Watt of other components such as RAM and storage (GPU is by far the bottleneck so the CPU is not running at 100%. Actual wattage might even be lower. But this is a ballpark estimate. So total 500W.
- I've tested a bit and it seems that a block of 50 sentences takes around 23 seconds, and Kreuze has about 120k sentences. This gives 120k/50 blocks * 23 seconds/block / 3600 seconds per hour = about 15.8 hours of total runtime.
- electric energy price for consumers is just under €0,30 per kWh in the Netherlands
- So 500W *15.8h / 1k = 7.9 kWh -> €2,37

This is great! I'd do it like this even if the GPU were ten times less efficient. No current big tech api approach comes close to being this cheap.

**Comment by tom9358** [comment]
I realized that I might try to do this with my own NLLB model first, hoping that an LLM might not be needed at all. So, I tried backtranslating with the NLLB model I have now. I did this at epochs 4, 8 and 12 (all have similar gos->nld validation scores but training score kept increasing). Based on some heuristics checks detecting untranslated Gronings words and some manual inspection, it was immediately clear that it actually performed really badly. I do not think those sentences would be a valuable quality synthetic data addition. I'll upload the data, though. Would've been nice, since the entire generation only took like 15 minutes. Anyway, considering the cost and time to do it with Qwen/Gemma, and the much better preliminary results with those models, I'm gonna go ahead and continue with the LLM investigation.

**Comment by tom9358** [comment]
Chronological documentation of my research process:

## Part 1: Qwen experiments

The local `Qwen3.6-35B-A3B-GGUF:Q8_K_XL` model was tested through a local OpenAI-compatible llama server. The first small probe was mixed: some translations were good, but it made errors with pronouns, names, and dialect words. It seemed better than my NLLB model.

I looked at if few-shot prompting would help. I made sure to select rather long sentences as examples, so the model would get some syntactical information as well. I took a fixed 200-sentence Tatoeba set (so I had "gold standards" and I could calculate bleu and chrf scores). I also checked if using or leaving out a system prompt would help, and I varied the prompting a bit and tried nonzero temperature once:

| Prompt setup | BLEU | chrF | Exact |
|---|---:|---:|---:|
| 0-shot with system prompt | 59.33 | 75.03 | 64/200 |
| 5-shot, one user prompt, with system | 64.34 | 79.18 | 81/200 |
| 10-shot, one user prompt, with system | 64.16 | 79.56 | 81/200 |
| 5-shot, one user prompt, no system | 64.21 | 79.05 | 80/200 |
| **10-shot, one user prompt, no system** | **66.40** | **80.04** | **88/200** |
| 10-shot requesting translations be "natural, idiomatic" | 66.05 | 79.73 | 88/200 |
| 20-shot with the same extra wording | 64.76 | 79.55 | 79/200 |
| 10-shot, temperature 0.3 | 64.81 | 78.96 | 87/200 |

The best Qwen setup used one user message, ten examples, no system prompt,
and temperature 0. The extra wording and additional examples did not help.
Then I looked into whether Gemma might be any better.

## Part 2: Gemma experiments

The local `gemma-4-31B-it-GGUF:Q8_K_XL` model was tested under the same general conditions.
In zero-shot mode it tended to ignore the instruction to only return the translation and it kept giving explanations and alternative translations instead of only a translation. Few-shot examples fortunately caused it to follow the desired output format perfectly.

Super interestingly, being already positively surprised by Qwen, I was blown away by Gemma's performance when I saw the numbers side by side. It is MUCH better than Qwen!

| Prompt setup | Qwen BLEU | Qwen chrF | Qwen exact | Gemma BLEU | Gemma chrF | Gemma exact |
|---|---:|---:|---:|---:|---:|---:|
| 0-shot, with system | 59.33 | 75.03 | 64/200 | 56.80 | 72.96 | 65/200 |
| 5-shot, with system | 64.34 | 79.18 | 81/200 | 69.07 | 80.39 | 95/200 |
| 5-shot, no system | 64.21 | 79.05 | 80/200 | 68.60 | 80.12 | 96/200 |
| 10-shot, with system | 64.16 | 79.56 | 81/200 | 68.84 | 80.57 | 99/200 |
| **10-shot, no system** | **66.40** | **80.04** | **88/200** | **68.97** | **80.58** | **101/200** |

These are ceteris-paribus results swapping out only the model.

I did not look into 20-shot prompting or T=0.3 with Gemma anymore, nor did I adapt the prompt, as I think this was entering the territory of diminishing returns. I might be wrong, especially about the N-shot examples. But I was satisfied enough with these results to continue here.
I was dumb and saved the tatoeba testing outputs to `/tmp/` and consequently lost them. A rookie mistake, but the scripts and models are here and easy to replicate.

Some Gemma errors were genuine lexical or idiomatic errors, for example
misinterpreting words such as `brogge`, `schieterg`, `liepen`. This is a limitation I don't think we can solve without big engineering steps (e.g. automatically adding dictionary results to each prompt. and that might require lemmatizing, and might backfire if the dictionary results are not applicable.. I'm sure it can be done, but I'm not sure how good the value/effort ratio would be)
- Gold standard has limitations; Gronings "ie" and "joe" can mean either “u” or “jullie” and "joe" can be both subject and object. "heur" can be translated to "hun", "hen" or "haar" and some contexts allow all of these, Dutch also sometimes has interchangeable synonyms e.g. "la"/"lade", "jij/je", "goede/goeie";
- There are lots of other cases where the LLM might generate an acceptable translation that differs from the gold standard.

The scores therefore might _underestimate_ actual quality a little bit!

The conclusion of going for 10-shot Gemma without system prompt has been made, in any case!

**Comment by tom9358** [comment]
## Some thoughts that I wanna defer to the comment section of this issue for now:

The following might also be in diminishing returns territory, but I'm thinking it might be possible to improve the n-shot examples. The Qwen 20-shot experiment seemed to point at that raising n is not effective. But perhaps we could still get better results as follows:
- see which words are most common in the sentences that got translated worst, that don't occur as much in all other sentences (or find another way to get rid of words that are just super common)
- make a selection of sentence pairs from tatoeba that contain these words and add them to the prompt
- this should lead to these words being translated more successfully. what the impact is on translation quality as a whole, remains a question. we should probably then also split the 200 sentences so we can see if quality on sentences without these words deteriorates. I can imagine some downsides to doing this, but idk better just test it.

Before diving down that rabbit hole, I'd much rather do the 120k sentence generation and retrain the NLLB model and see how it performs.

(and while we're at it, it would be good to revisit the 1.3B model as well. A long time ago, I concluded that both models performed the same because they were just severely undertrained, not enough data. That might still be the case, but who knows. maybe a 10x of the training data changes things?)

**Comment by tom9358** [comment]
## Is translating big blocks feasible?

I tested on consecutive Kreuze sentences sentences, because I hoped if the LLM has a larger story, it would be easier to infer the right translations of Gronings words, anaphora, ambiguous words, multiple meanings and such from this extra context. I first tested if it could adhere to the format required for this more complex instruction at all:

As a quantitative batching test, eight blocks of 50 randomly grouped Tatoeba
sentences were translated in eight queries:

- 400/400 output lines were aligned;
- no fallback was needed;
- BLEU: 73.34;
- chrF: 83.72;
- exact matches: 193/400.

Note these sentences were randomly grouped, so this cannot reflect the possible advantages the context of larger passages might have.

I did tests with blocks of Kreuze text as well to ensure that there too, the output lines would be aligned with the input lines. I built in a fallback just in case, so that in case the output was not exactly 50 sentences, it was discarded and reprompted with two prompts of 25 sentences. If that failed, fallback to 5 sentences, if that failed to output 5 sentences, fallback to prompts with individual sentences. This way we wouldn't ever lose 49 sentences because a single one was causing trouble.

Context helped with references and ambiguous words. In a passage about water plants, for example, "kreuze" was correctly translated as "kroos" when the surrounding sentences supplied the context.
Without this context, it tended to become other similar-looking but unrelated words such as "kruis".

Total, que this shows that we can safely go ahead and translate big blocks! Adelanteeuh¡!¡!

## 5. The real work! More context by translating big blocks. Full Gemma generation

I almost wrote FULL BRIDGE RECTIFIER! I'm going a bit haywire from writing so much documentation fjdksaljfdklsajgrsjl

The full Kreuze collection was generated with Gemma in blocks of 50 consecutive sentences with fallbacks in case the in- and output were not the same number of sentences (fallback order 50, 25, 5, 1). Since this takes several hours, I also made sure the generator supports resuming an interrupted run (e.g. power outage or any other kind of sudden process kill).

The next day, tadaah! The resulting corpus contains 104,447 Gronings-Dutch pairs from 103 source
documents. It required 2,268 local API queries.
This is lower than the 120k lines in the input json, because:
- there were about 18k duplicate lines
- there were 746 lines of more than 200 characters. This was an artifact of copying code from the NLLB training notebook, which slipped into the generation code, oops! Maybe these should be left in, given that such long sentences could contain complex syntactic information, but many of these are poems or lyrics without punctuation, or simply wrongly segmented texts. (<- see here an example of a sentence of exactly 200 characters.)

Good that we built in the fallbacks, because we get:

| Effective block size | Rows |
|---:|---:|
| 50 | 101,800 |
| 25 | 2,125 |
| 5 | 470 |
| 1 | 5 |
| final block of 47 | 47 |

The output files, which I've pushed just now, are:

```text
data/kreuze/kreuze_synthetic_gemma50.csv
data/kreuze/kreuze_synthetic_gemma50.jsonl
```

They look not perfect, but quite good. There is no gold standard for this data, of course. But they look definitely good enough to be used as extra training data for a translation model.

## Final chosen setup

The current synthetic-data setup is:

- model: `hf.co/unsloth/gemma-4-31B-it-GGUF:Q8_K_XL`;
- Gronings to Dutch;
- one user prompt, no system prompt;
- ten fixed long Gronings -> Dutch examples from the Tatoeba training data;
- temperature `0`;
- seed `9358`;
- thinking disabled;
- 50 consecutive Kreuze sentences per request (with fallback);
- output to both CSV and JSONL because it seemed practical (might seem excessive but the csv is a nice format for training and the jsonl also documents the source of each sentence)

The local source copy used by the generators is
`sidetracks/kreuze/sentences_kreuze.json`. I plan to now use this Gemma corpus for further NLLB-training experiments.
It remains synthetic data and should not be treated as gold reference data without further quality filtering! It is NOT perfect.

**Comment by michaelmanly** [comment]
@tom9358 
the fallback + resume setup is nice. once you’re doing 100k+ rows, the annoying part becomes making sure a long generation run can fail/restart without babysitting it.

if you’ve got the exact generation command, I can give you a bounded run command for it with resume/logging/cleanup built around the workload.

**Comment by tom9358** [comment]
> [@tom9358](https://github.com/tom9358) the fallback + resume setup is nice. once you’re doing 100k+ rows, the annoying part becomes making sure a long generation run can fail/restart without babysitting it.
> 
> if you’ve got the exact generation command, I can give you a bounded run command for it with resume/logging/cleanup built around the workload.

Hi! Thanks for the message. Too late, it's already finished, as my previous comment says! I dont blame you for not noticing because I wrote with myself as the only reader in mind. Probably not easy to read for others xD

Generation finished without any hiccups, but I am curious what you mean with a "bounded run command". I'm not sure what it is. Basically I just ran `uv run sidetracks/kreuze/generate_kreuze_gemma.py --output data/kreuze/kreuze_synthetic_gemma50.csv`. Of course it could have failed for many reasons, so I'm interested to see what you had in mind!
 And I wonder, how did you even come across this issue? It's quite a niche project I think 😄

**Comment by tom9358** [comment]
# Super cool findings:
For big LLM-written documentation word barf, see https://github.com/tom9358/nllb-tryout/blob/main/data/kreuze/training_plan.md

I first did an experiment using only tatoeba's Dutch-Gronings data (no multilingual supporting languages yet). I wanted to know if the synthetic data improves translation quality. The table below does not contain results of all stages of the experiment.

| Validation | Direction | Metric | 600M Tatoeba-only | 600M synthetic + clean | 1.3B Tatoeba-only | 1.3B synthetic + clean |
|---|---|---|---:|---:|---:|---:|
| Tatoeba | Dutch→Gronings | BLEU | 57.16 | 55.81 | 57.17 | **60.93** |
| Tatoeba | Dutch→Gronings | chrF | 76.72 | 77.04 | 77.48 | **81.01** |
| Tatoeba | Gronings→Dutch | BLEU | 71.96 | 73.99 | 77.14 | **80.63** |
| Tatoeba | Gronings→Dutch | chrF | 83.33 | 84.04 | 86.13 | **88.23** |
| Kreuze | Dutch→Gronings | BLEU | 34.09 | 48.21 | 34.12 | **49.89** |
| Kreuze | Dutch→Gronings | chrF | 61.29 | 72.79 | 61.14 | **74.41** |
| Kreuze | Gronings→Dutch | BLEU | 47.44 | 63.41 | 49.64 | **66.32** |
| Kreuze | Gronings→Dutch | chrF | 67.63 | 78.81 | 68.88 | **80.77** |

So, answer: It greatly improves Kreuze generalization but initially causes a Tatoeba-domain drop. (The leftmost score column. A kind of OOD effect I suppose)

Followup question: Can adding one clean Tatoeba pass after synthetic training recover Tatoeba performance while retaining synthetic knowledge?

Answer: Yes!!! Tatoeba performance recovers while most synthetic-domain gains remain. 😎 (second score col)

Another question I had: Can the 1.3B model generalize better from the more varied (and quite noisy) synthetic data?

Answer: Yes!! 1.3B synthetic plus clean beats equal-compute Tatoeba-only on all eight metrics. (the other two score cols)

---

\<ramble>
Before, I trained multilingually and then I let each language pair have the same number of data points (if enough data was available) as the Dutch-Gronings pair. Since that pair now has way more data, this would blow up compute if we kept English, German and Spanish all in the training data. I decided to solve this by reducing the other languages' samples and making 50% of all training samples Dutch-Gronings pairs. This shrinks the size of an epoch from 686,563 to 230,712 samples. Haven't looked into it in much detail how it affects multilingual translation quality. Haven't really thought this through that well yet hehe. But I need some way to keep it computable.
\</ramble>

Anyway, multilingual finetuning will be the next step! In particular, I'll train with Dutch, Gronings, English, German, Spanish, and the Kreuze synthetic pairs, followed by a clean multilingual pass. I wanna know if this prevents catastrophic forgetting of the translation capacity in the other languages, and more importantly, if it is better than a multilingually trained model without the synthetic data.

**Comment by tom9358** [comment]
## Multilingual results

Using the sampling idea above to reduce excessive computation on other language pairs than Dutch-Gronings, i.e.
50% Dutch-Gronings, 50% distributed across the other nine language pairs using temperature balancing, which makes 230,712 samples per epoch instead of 686,563, we find the following (summarized, scores in `training_plan.md`) for the 600M model that both Dutch->Gronings and Gronings->Dutch _tatoeba_ performance _improve_ a bit when going multilingual! Multilingual = GOOD

Then I repeated this with the 1.3B model. It had the same results, and scored marginally better than the 600M model.

The exact two-staged training recipe I followed:

### First stage: Train for eight epochs with 230,712 samples per epoch:

| Data group | Samples/epoch | Share (%) |
|---|---:|---:|
| Synthetic Kreuze Dutch–Gronings | 102,318 | 44.4 |
| Clean Tatoeba Dutch–Gronings | 13,038 | 5.7 |
| Other multilingual Tatoeba pairs | 115,356 | 50 |

Supporting languages distributed over nine language pairs with temperature balancing at `T=5`.

### Second stage:

A "clean finish" that should counteract mistakes that the model learned from the synthetic Kreuze data. This is only a single pass of all tatoeba Dutch-Gronings data supplemented 50/50 by the T-balanced support language data, i.e., 13038 x2 samples in total.

**Comment by tom9358** [comment]
## How much should we train?

My next question was: more epochs, more better?

For a translation model, memorization of good quality training data is not necessarily a problem, although we do want the model to generalize its learnings of course. So, in principle, we can keep training until validation scores plateau.
So, I ran two trainings the 1.3B model for both 12 epochs, one with only Tatoeba data, and one with Tatoeba+Kreuze.

Basically scores do go up until epoch 8, but then plateau (NLD->GOS tatoeba validation scores plateau around BLEU ~70 and chrF ~85, so quite a bit higher than the 2-epoch scores of ~60 and ~80 respectively).

Now that we've found a way to train the model multilingually without making the total training data volume dependent on the nuber of other languages, I'm also thinking about adding a handful more languages. Although I wouldn't be surprised if it gives diminishing returns. We already have 20 directions in total (5 languages so 5*4=20), after all. Maybe French would be worthwhile, as it also has ~1000 Gronings pairs.

I could also still try to double the clean tatoeba finish (swapping the translation direction of each sentence) and see what happens.

**Comment by tom9358** [comment]
> Now that we've found a way to train the model multilingually without making the total training data volume dependent on the nuber of other languages, I'm also thinking about adding a handful more languages. Although I wouldn't be surprised if it gives diminishing returns. We already have 20 directions in total (5 languages so 5*4=20), after all. Maybe French would be worthwhile, as it also has ~1000 Gronings pairs.

See issue #35 

> I could also still try to double the clean tatoeba finish (swapping the translation direction of each sentence) and see what happens.

See issue #37

**Comment by tom9358** [comment]
This issue including a lot of its scope creep is finished! Data generated successfully (although I wasnt super rigorous in LLM selection; I tested only two. But one already performed really quite well so it seemed fine for now. no ragrets)! Also, finetuning with it done successfully and performance improved quite a bit! To be continued.
